src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,289 @@
+            // see "Microsoft.CodeAnalysis.Editor.Implementation.LanguageClient.AlwaysActivateInProcLanguageClient" for the list of content types we register for.
+            // At this time the expected list is C#, Visual Basic, and F#
+            if (response is null)
+            {
+                return null;
+            }
+
+            var responseBody = response.Value.response.ToObject<DocumentSymbol[]>();
+            // It would be a bug in the LSP server implementation if we get back a null result here.
+            Assumes.NotNull(responseBody);
+
+            var model = DocumentOutlineHelper.CreateDocumentSymbolDataModel(responseBody, response.Value.snapshot);
+
+            _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, CaretPositionOfNodeToSelect: null, ExpansionOption: null, DataUpdated: true));
+
+            return model;
+        }
+
+        private void EnqueueFilter(string? newText)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(newText, CaretPositionOfNodeToSelect: null, ExpansionOption: null, DataUpdated: false));
+
+        public void EnqueueSelectTreeNode(CaretPosition caretPoint)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, caretPoint, ExpansionOption: null, DataUpdated: false));
+
+        public void EnqueueExpandOrCollapse(ExpansionOption option)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, CaretPositionOfNodeToSelect: null, option, DataUpdated: false));
+
+        private async ValueTask UpdateViewModelStateAsync(ImmutableSegmentedList<ViewModelStateDataChange> viewModelStateData, CancellationToken cancellationToken)
+        {
+            _threadingContext.ThrowIfNotOnBackgroundThread();
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var model = await _documentSymbolQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            var searchText = viewModelStateData.SelectLastNonNullOrDefault(static x => x.SearchText);
+            var position = viewModelStateData.SelectLastNonNullOrDefault(static x => x.CaretPositionOfNodeToSelect);
+            var expansion = viewModelStateData.SelectLastNonNullOrDefault(static x => x.ExpansionOption);
+            var dataUpdated = viewModelStateData.Any(static x => x.DataUpdated);
+
+            // These updates always require a valid model to perform
+            if (model is not null)
+            {
+                if (string.IsNullOrEmpty(searchText) && dataUpdated)
+                {
+                    var documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(model.DocumentSymbolData);
+                    ApplyExpansionStateToNewItems(documentSymbolViewModelItems, DocumentSymbolViewModelItems);
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
```

Comment:
trying to figure this logic out.  So if there's no search text, but data is updated, then we go make the items, and apply the old to the new.  but what if htere *is* search text and data is updated.  Do we just lose the updated model items?

Reply:
We always use the data from the latest snapshow this logic is to determine if we need to worry about expand / collapse. If we are searching we always expand the visible items so you can see them